### PR TITLE
Fixed a crash when loading a JAR library that exports tools not derived from AddTool

### PR DESCRIPTION
**Pull request by flarn2006**
I tried importing a JAR library that exports a tool which extends `Tool` rather than `AddTool`, and it gave me a Java error pointing to `LogisimFile.addLibrary`. I checked the code, and it looks like that method is coded to make the Name attribute read-only for `AddTool`s that are imported, but in doing so, it casts it to `AddTool` right away without first checking to ensure it actually *is* an `AddTool`.

**Note: I did not test this change**, as it wouldn't compile for me. So while this should work as far as I can tell, it's possible that this might not actually fix the issue, or might cause a different issue. It should compile fine for you, however—at first there was an error in the change I made, but after fixing that, it gave me a different error, and that error appears even when I try to compile the unmodified code. ([Here's the build scan.](https://scans.gradle.com/s/lxdpq63ezu5cq)) So I know I didn't cause the problem that prevents me from compiling it.

### CODE REVIEW

**Comment by flarn2006** [comment]
Update: I noticed that the error I was getting was actually after the compiling stage, so I still had the compiled `.class` file with the code I modified. I replaced the one in a prebuilt .jar with that one, and it appears to work, though so far I've only tested it with a trivial "tool" that only defines `getDisplayName` and `getDescription` and thus does nothing.

**Comment by maehne** [comment]
Thanks for the PR! However, please redo it against the `develop` branch instead of `master`. From my perspective, it looks sensible. However, @BFH-ktt1 should als have a look.

File: src/main/java/com/cburch/logisim/file/LogisimFile.java [hunk 1]
```diff
@@ -253,10 +253,12 @@ public void addVhdlContent(VhdlContent content, int index) {
   public void addLibrary(Library lib) {
     if (!lib.getName().equals(BaseLibrary._ID)) {
       for (final var tool : lib.getTools()) {
-        final var tool1 = (AddTool) tool;
-        final var atrs = tool1.getAttributeSet();
-        for (Attribute<?> attr : atrs.getAttributes()) {
-          if (attr == CircuitAttributes.NAME_ATTR) atrs.setReadOnly(attr, true);
+        if (tool instanceof AddTool) {
+          final var tool1 = (AddTool) tool;
```
Comment (MarcinOrlowski) [inline src/main/java/com/cburch/logisim/file/LogisimFile.java:257]: Can you please rename `tool1` variable to something more descriptive, i.e. `addTool`?

File: src/main/java/com/cburch/logisim/file/LogisimFile.java [hunk 2]
```diff
@@ -253,10 +253,12 @@ public void addVhdlContent(VhdlContent content, int index) {
...
+          final var atrs = tool1.getAttributeSet();
+          for (Attribute<?> attr : atrs.getAttributes()) {
```
Comment (MarcinOrlowski) [inline src/main/java/com/cburch/logisim/file/LogisimFile.java:259]: can we have `var` instead of `Attribute<?>`?

**Review by MarcinOrlowski — CHANGES_REQUESTED** [review]

File: src/main/java/com/cburch/logisim/file/LogisimFile.java (hunk 1 above)
Comment (MarcinOrlowski) [reply src/main/java/com/cburch/logisim/file/LogisimFile.java:257]: Also can you please replace `tool instanceof AddTool` with `tool instanceof AddTool addTool` and then remove the `(AddTool)` casting line below so it reads: `final var tool1 = addTool;`?

**Review by maehne — COMMENTED** [review]
@flarn2006: Could you please address the changes requested by @MarcinOrlowski ?

**Comment by flarn2006** [comment]
Oh shit, I completely forgot! Remind me to take a look at this tomorrow.

File: src/main/java/com/cburch/logisim/file/LogisimFile.java (hunk 1 above)
Comment (flarn2006) [reply src/main/java/com/cburch/logisim/file/LogisimFile.java:257]: I'm confused; do you want me to change `tool1` or not? In your first comment you said to make it something more descriptive, but then in your second comment you said to change it to something containing `tool1`.

**Comment by flarn2006** [comment]
I did the changes in a new branch based on `develop`; how do I incorporate those changes into this pull request? Should I force push?

File: src/main/java/com/cburch/logisim/file/LogisimFile.java (hunk 1 above)
Comment (MarcinOrlowski) [reply src/main/java/com/cburch/logisim/file/LogisimFile.java:257]: I asked to renamed variable `tool1` to i.e. `addTool` and also, I asked to replace

```java
        if (tool instanceof AddTool) {
          final var tool1 = (AddTool) tool;
```

to 

```java
        if (tool instanceof AddTool tool1 {
```

NOTE: I left `tool1` for reference only to not complicate my example. After both changes made there should be no `tool1` variable.

**Comment by MarcinOrlowski** [comment]
> how do I incorporate those changes into this pull request?

@flarn2006  I'd first edit this PR and change target branch here:

x


And if you already merged your working branch with `develop/` then simply push the update to this PR and that should be all needed.

**Comment by flarn2006** [comment]
How's that?

File: build.gradle.kts [hunk 3]
```diff
@@ -54,6 +54,14 @@ dependencies {
   testImplementation("org.mockito:mockito-junit-jupiter:4.2.0")
 }
 
+sonarqube {
```
Comment (MarcinOrlowski) [inline build.gradle.kts:57]: I do not find this in scope of this PR. Please remove.

**Review by MarcinOrlowski — CHANGES_REQUESTED** [review]

**Comment by flarn2006** [comment]
Those were from merge conflicts that I guess I messed up. I'll take a look.

**Comment by flarn2006** [comment]
That better?

**Comment by MarcinOrlowski** [comment]
Yes, that's now all in scope. Thanks.

**Review by maehne — APPROVED** [review]
Looks good to me! Thanks @flarn2006 and @MarcinOrlowski!

**Review by MarcinOrlowski — APPROVED** [review]

**Pull request merged**